Fix import of CAPI RSA keys.

justremotephone opened the pull request:
BigIntegers must be positive for RSAPublicKeySpec. Therefore add a leading zero byte in the case the most significant bit is set.

I was not able to import RSA public keys created on a Windows desktop client into an Android key. After some debugging I found that Modulus and Exponent are correctly received at the Android site but that the native key creation fails. I am not sure, but may be that your tires around KeyFormatter and removing/adding leading zeros are now obsolete.

I hope will merge this...

Thanks for your work!

Regards,
Thomas.

AArnott commented:
Please include a test change with this where the test will fail without your product change and passes with the change. I suspect you are correct, and it's a subtle enough behavior that I want to defend against future regressions by asking for a test.

justremotephone commented:
Hello,

I have added a test case and while testing I have found out furhter details:

The problem only appears on Android 4.0 and is fixed with Android 4.1. (I have tested this on emulator and on a real device)

I  have also updated my fix with some code improvements, but it is working like the first version.

I hope this helps!

And please do not forget, this is a BUG ON ANDROID 4.0

AArnott commented:
Thanks! I will verify and merge this next time I'm on my Xamarin computer.

EgorBo commented:
@AArnott it would be nice to have this pull request in a nuget package.

AArnott commented:
I'm trying to reproduce the problem and apply your fix to verify now.

AArnott commented:
The fix looks effective. Thank you so much for your contribution.
